src/seedu/addressbook/data/person/Contact.java

```diff
@@ -0,0 +1,48 @@
+package seedu.addressbook.data.person;
+
+import seedu.addressbook.data.exception.IllegalValueException;
+
+/**
+ * represents a parent of phone, email, address
+ */
+public abstract class Contact {
+	public final String value;
+    private boolean isPrivate;
+    
+    public Contact(String contact, boolean isPrivate, String messageConstraints, String validationRegex) throws IllegalValueException {
+        this.isPrivate = isPrivate;
+        String trimmedContact = contact.trim();
+        if (!isValidContact(trimmedContact, validationRegex)) {
+            throw new IllegalValueException(messageConstraints);
+        }
+        this.value = trimmedContact;
+    }
+    /**
+    * Checks if a given string is a valid person contact.
+     * @param trimmedContact 
+    */ 
+    public static boolean isValidContact(String test, String validationRegex) {
+        return test.matches(validationRegex);
+    }
+    	
+    @Override
+    public String toString() {
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof Contact // instanceof handles nulls
+                && this.value.equals(((Contact) other).value)); // state check
```

Comment:
Note that now an Email and Address having the same value will be considered equal even if they are of different types.